### DISCUSSION TITLE
Create API keys in Supabase for roles other than "anon" and "service"

### DISCUSSION THREAD

**Discussion by joaoguidev** [supabase/supabase#20983] [Questions] [answered]
Hello.

I created a new role but I get only an empty array [] when .select('*') from a table were I defined the RLS to allow select for the new role.
Should I grant something to this new role ?

******* What I did was ******* 

I created a new role named my_new_role on the supabase dashboard.

Than I mint a new jwt to be used with my new role following this short [tutorial](https://dev.to/acetrondi/how-to-create-api-keys-in-supabase-for-roles-other-than-anon-and-service-1j7f).

Next I created the supabase client like this and tried to query the table :

```
import { createClient } from "@supabase/supabase-js"

export async function createSupabaseClient({ request, context }) {
   const options = {
      global: {
        headers: { 'Authorization': 'Bearer ' + context.env.SUPABASE_MY_NEW_ROLE_KEY }
      },
    }
    const supabase = createClient(context.env.SUPABASE_URL, context.env.SUPABASE_ANON_KEY, options )
    const { data, error } = await supabase.from("test_table").select("*")
    console.log('Data',data);
    console.log('error',error);
   }
}
```

**Comment by GaryAustin1** [comment, marked as answer]
Towards the middle to end of this: https://github.com/orgs/supabase/discussions/11948 shows more.
You at least need to assign the role to authenticator role.

**Reply by joaoguidev** [reply]
Hi.

On this article they are doing somithing a lot more complex than what I want. They are assossiating the auth of users whith db roles from what I understood.

What I want is relatively simple, effectivelly I want to have a second anon key to query supabase.

**Reply by GaryAustin1** [reply]
You are seemingly doing the bulk of what is there as far as wanting PostgREST to know about your role and be able to assign it from the jwt.   You are just not using the auth.users table part.  You have to assign the role to authenticator at a minimum.

**Reply by GaryAustin1** [reply]
Also when you set the authorization header you need to use 'bearer jwt'.

**Reply by joaoguidev** [reply]
Hello.

If i do `GRANT my_new_role TO anon;` It works but I have a problem. 

I have a table "test_table" with RLS authorizing SELECT only to the role my_new_role. 

The problem is that after the GRANT mentioned I can select from this table either with the anon role or my_new_role. Both works dispite the RLS is set only to my_new_role..

************** ANON ROLE ************
 ` const supabase =  await createClient(context.env.SUPABASE_URL, context.env.SUPABASE_ANON_KEY)`  

************** my_new_role ROLE ************

```
const options = {
      global: {
        headers: { 'Authorization': 'Bearer ' + context.env.SUPABASE_MY_NEW_ROLE_KEY }
      },
    }
    const supabase = createClient(context.env.SUPABASE_URL, context.env.SUPABASE_ANON_KEY, options )

```

If I `REVOKE my_new_role FROM anon;` and query using   my_new_role ROLE mentioned above I get the error: `{
  code: '42501',
  details: null,
  hint: null,
  message: 'permission denied to set role "my_new_role"'
}`

**Reply by GaryAustin1** [reply]
Did you assign your role to authenticator role like shown in my link and mentioned above?
![image](https://github.com/supabase/supabase/assets/54564956/b8f3eab5-68fa-43dd-8978-aea22c57f52c)

**Reply by joaoguidev** [reply]
Ups. My mistake. 

Because I was trying to make a "second anon" I was granting to the anon instead of authenticator.

It worked!

Sorry I have been tweking this for hours straight.

Thank you.

**Reply by GaryAustin1** [reply]
You can assign anon to your new role also and then it will work wherever anon works.

**Reply by joaoguidev** [reply]
Awesome! Thank you so much.
